zetaclient/chains/sui/client/client.go

```diff
@@ -250,40 +251,69 @@ func (c *Client) SuiExecuteTransactionBlock(
 	return parseRPCResponse[models.SuiTransactionBlockResponse]([]byte(resString))
 }
 
-// GetSuiCoinObjectRef returns the latest SUI coin object reference for given owner address
-// Note: the SUI object may change over time, so we need to get the latest object
-func (c *Client) GetSuiCoinObjectRef(ctx context.Context, owner string) (suiptb.ObjectRef, error) {
+// GetSuiCoinObjectRef returns the latest SUI coin object reference for given owner address and wanted balance
+// Note: the SUI coin object may change over time, so we need to use the up-to-date SUI coin object
+func (c *Client) GetSuiCoinObjectRef(ctx context.Context, owner string, wantBalance uint64) (suiptb.ObjectRef, error) {
 	coins, err := c.SuiXGetCoins(ctx, models.SuiXGetCoinsRequest{
 		Owner:    owner,
 		CoinType: string(zetasui.SUI),
 	})
-	if err != nil {
+
+	switch {
+	case err != nil:
 		return suiptb.ObjectRef{}, errors.Wrap(err, "unable to get TSS coins")
+	case len(coins.Data) == 0:
+		return suiptb.ObjectRef{}, errors.New("no SUI coin found")
 	}
 
+	// sort coins by object ID to make the result deterministic across observres
+	sort.SliceStable(coins.Data, func(i, j int) bool {
+		return coins.Data[i].CoinObjectId < coins.Data[j].CoinObjectId
+	})
+
 	var (
-		suiCoin        *models.CoinData
-		suiCoinVersion uint64
+		suiCoin           *models.CoinData
+		suiCoinVersion    uint64
+		suiCoinVersionStr string
 	)
 
-	// locate the latest version of SUI coin object of given owner
+	// get the latest version from owned SUI coin objects
 	for _, coin := range coins.Data {
 		if !zetasui.IsSUICoinType(zetasui.CoinType(coin.CoinType)) {
-			continue
+			return suiptb.ObjectRef{}, fmt.Errorf("invalid coin type: %s", coin.CoinType)
 		}
 
 		version, err := strconv.ParseUint(coin.Version, 10, 64)
 		if err != nil {
-			return suiptb.ObjectRef{}, errors.Wrapf(err, "failed to parse SUI coin version %s", coin.Version)
+			return suiptb.ObjectRef{}, errors.Wrapf(err, "invalid version %s", coin.Version)
 		}
 
-		if version > suiCoinVersion {
-			suiCoin = &coin
+		if version >= suiCoinVersion {
```

Comment:
What's the purpose of finding the highest version, is it a requirement? Should be explain in the comments if so

Reply:
Not required. And I realized that using multiple SUI coin object is a better: https://github.com/zeta-chain/node/pull/3877/commits/ca5204b8a1aa295c450211f22ddaa90b6b00ac1c
Using multiple coin objects is similar to Bitcoin UTXO consolidation.